### PR TITLE
[imgui] WIP: allow split outputs

### PR DESCRIPTION
**Pull request by SomeoneSerge**
A minimal set of changes required to allow "splayed layouts", installation of different components into independent prefixes, in https://github.com/NixOS/nixpkgs/pull/303682. This is currently pretty coarse and oblivious of vcpkg' styles and practices, but I can adapt this further after we've confirmed this might be eventually merged

### CODE REVIEW

**Comment by SomeoneSerge** [comment]
@microsoft-github-policy-service agree

**Comment by BillyONeal** [comment]
It seems like you're looking for feedback on whether we would take this in the first place. I'm not sure, I think we would need:

(1) A more concrete description of the problem that this fixes for vcpkg customers
(2) The build to not be broken

before we would take this.

**Comment by SomeoneSerge** [comment]
> It seems like you're looking for feedback on whether we would take this in the first place

Hi, yes, thank you, that's exactly what I meant to ask.

> IMO the change indicates the vcpkg port becoming a source for a CMake build system over imgui. This is not the role of the vcpkg registry. Ideally, such vendored build systems should not need to change with version updates in the vcpkg registry.

I'll start by addressing this one. There was no intention to impose on Vcpkg the responsibility of serving as a source of truth or anything of that sort. The Nixpkgs change has motivated the present PR, but it's independent: it takes the risk of using Vcpkg' cmake target names as a reference; it's based on the assumption that the Vcpkg cmake rules are the next obvious choice an average user might make, other than vendoring imgui; I could've as well written a custom CMakeLists.txt for Nixpkgs that would try to stay compatible with Vcpkg and the `find_package` level, but at that point it just makes sense to reuse the existing rules; if Vcpkg should change in an incompatible way, it's my responsibility to make whatever relevant changes in Nixpkgs.

> (1) A more concrete description of the problem that this fixes for vcpkg customers

I'm not sure how common is it to `cmake --install` dependencies obtained via Vcpkg into external prefixes, but I see that many other "recipes" in Vcpkg are written specifically in a way that accommodates relocatability and splitting of components: just code-searching e.g. for [`CMAKE_INSTALL_INCLUDEDIR`](https://github.com/search?q=repo%3Amicrosoft%2Fvcpkg+CMAKE_INSTALL_INCLUDEDIR&type=code) I see a number of `.diff`s and full CMakeLists files, e.g. https://github.com/microsoft/vcpkg/blob/94a50e8c1357d423840e22451e122f56d62b5c20/ports/pcre2/fix-cmake.patch#L17-L20

So I could instead say that this change seeks to improve on consistency and bring the imgui module to the same standard as other parts of Vcpkg. I don't know what exactly that standard is and what range of use-cases it's addressing, which is where I'm hoping to hear your advice.

The specific changes I pushed are the ones that allowed me to simply reuse vcpkg' cmake rules in Nixpkgs (there the constraint was that we might install the headers and cmake targets into a separate prefix `/aaaa/imgui-dev/`, the `.a`-archives into `/bbbb/imgui-static`, and shared objects into `/cccc/imgui-dev`, and we wish the generated cmake target to still correctly account for all these paths)

**Comment by SomeoneSerge** [comment]
> (2) The build to not be broken

Oh yes about that, I couldn't at first figure out how to access the build logs, and I think now they got garbage-collected? This has been verified on `{x86_64,aarch64}-{linux,darwin}`, but only in simplest configurations (e.g., on linux, glfw, no svg, no freetype), not across the full matrix

File: ports/imgui/CMakeLists.txt
Comment (SomeoneSerge) [inline ports/imgui/CMakeLists.txt:1]: I needed the patches rebased on master, but this time around I got to look at the logs. Here's e.g. x64-linux:

<details>

```

2024-06-26T21:29:10.2494199Z CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:112 (message):
2024-06-26T21:29:10.2495945Z     Command failed: /mnt/vcpkg-ci/downloads/tools/ninja/1.10.2-linux/ninja -v
2024-06-26T21:29:10.2496726Z     Working Directory: /mnt/vcpkg-ci/b/imgui/x64-linux-rel/vcpkg-parallel-configure
...
error: building ogre:x64-linux failed with: CASCADED_DUE_TO_MISSING_DEPENDENCIES
  due to the following missing dependencies:
    imgui[core]:x64-linux
    imgui[freetype]:x64-linux
...
SUMMARY FOR x64-linux
  SUCCEEDED: 93
  BUILD_FAILED: 1
  CASCADED_DUE_TO_MISSING_DEPENDENCIES: 10
REGRESSIONS:
REGRESSION: imgui:x64-linux failed with BUILD_FAILED. If expected, add imgui:x64-linux=fail to /mnt/vss/_work/1/s/scripts/azure-pipelines/../ci.baseline.txt.

##[error]PowerShell wrote one or more lines to the standard error stream.
##[error]REGRESSIONS:
REGRESSION: imgui:x64-linux failed with BUILD_FAILED. If expected, add imgui:x64-linux=fail to /mnt/vss/_work/1/s/scripts/azure-pipelines/../ci.baseline.txt.
```

</details>

Can't seem to find the original error

**Comment by FrankXie05** [comment]
@SomeoneSerge 
You can download the CI error log from here：
https://dev.azure.com/vcpkg/public/_build/results?buildId=104404&view=artifacts&pathAsName=false&type=publishedArtifacts

image

eg: error logs from `x64-linux`:
```
CMake Error at CMakeLists.txt:181 (install):
  install FILES given no DESTINATION!


CMake Error at CMakeLists.txt:255 (install):
  install FILES given no DESTINATION!


CMake Error at CMakeLists.txt:263 (install):
  install FILES given no DESTINATION!


-- Configuring incomplete, errors occurred!

```
[failure logs for x64-linux (5).zip](https://github.com/user-attachments/files/16006582/failure.logs.for.x64-linux.5.zip)

File: ports/imgui/CMakeLists.txt [hunk 1]
```diff
@@ -8,13 +8,15 @@ if(APPLE)
     enable_language(OBJCXX)
 endif()
 
+include(GNUInstallDirs) # Defines CMAKE_INSTALL_INCLUDEDIR if not set
```
Comment (SomeoneSerge) [inline ports/imgui/CMakeLists.txt:11]: Similarly, I'm just repeating after existing vcpkg expressions: https://github.com/microsoft/vcpkg/blob/c247b81088dbe46c159be6157944d5db8bf76b4a/ports/tgc/CMakeLists.txt#L5C1-L13

**Comment by BillyONeal** [comment]
>I'm not sure how common is it to cmake --install dependencies obtained via Vcpkg into external prefixes, but I see that many other "recipes" in Vcpkg are written specifically in a way that accommodates relocatability and splitting of components: just code-searching e.g. for [CMAKE_INSTALL_INCLUDEDIR](https://github.com/search?q=repo%3Amicrosoft%2Fvcpkg+CMAKE_INSTALL_INCLUDEDIR&type=code) I see a number of .diffs and full CMakeLists files, e.g.

Those are about making `find_package` work at all for upstreams that don't exactly match vcpkg's conventions for installed layout.

>So I could instead say that this change seeks to improve on consistency and bring the imgui module to the same standard as other parts of Vcpkg. I don't know what exactly that standard is and what range of use-cases it's addressing, which is where I'm hoping to hear your advice.

Right, I'm not seeing what the consistency change is here. Like, what is the thing a vcpkg customer can not do today that they could do with this change that would matter to them? I mean something like "as a vcpkg customer I want to be able to write

    find_package(imgui A B C D E F G)

as recommended by upstream at <LINK>, but today with how vcpkg installs imgui that fails because <REASONS>.

If you are looking for the vcpkg maintainers to provide justification for the feature you're contributing, that's kind of backwards and I don't believe we are going to do that.

**Comment by SomeoneSerge** [comment]
As I said before, I install different components (headers, DSOs) into separate components:

> install the headers and cmake targets into a separate prefix /aaaa/imgui-dev/, the .a-archives into /bbbb/imgui-static, and shared objects into /cccc/imgui-dev, and we wish the generated cmake target to still correctly account for all these paths

Most of vcpkg' ports (accidentally?) support this use case because they use `$<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>` and `GNUInstallDirs`. With `$<INSTALL_INTERFACE:include>` the generated `imguiConfig.cmake` expects headers and DSOs in the same prefix:

```
imnodes>   Imported target "imgui::imgui" includes non-existent path                
imnodes>            
imnodes>     "/aaaa-imgui-1.90.6/include" # Editor's note: the correct path is `/bbbbb-imgui-1.90.6-dev/include`, supplied at build time via `CMAKE_INSTALL_INCLUDEDIR` and recognized by GNUInstallDirs but ignored by vcpkg' imgui port
```

...whereas with `CMAKE_INSTALL_INCLUDEDIR` the generated cmake-config accounts for the user-supplied installation directories. If supporting GNUInstallDirs is out of scope for Vcpkg, I suppose this MR may be closed

> as recommended by upstream at , but today with how vcpkg installs imgui that fails because

As I said before, upstream chooses not to distribute cmake or any other build info.

**Comment by BillyONeal** [comment]
> Most of vcpkg' ports (accidentally?) support this use case because they use `$<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>` and `GNUInstallDirs`. With `$<INSTALL_INTERFACE:include>` the generated `imguiConfig.cmake` expects headers and DSOs in the same prefix:
> 
> ```
> imnodes>   Imported target "imgui::imgui" includes non-existent path                
> imnodes>            
> imnodes>     "/aaaa-imgui-1.90.6/include" # Editor's note: the correct path is `/bbbbb-imgui-1.90.6-dev/include`, supplied at build time via `CMAKE_INSTALL_INCLUDEDIR` and recognized by GNUInstallDirs but ignored by vcpkg' imgui port
> ```
> 
> ...whereas with `CMAKE_INSTALL_INCLUDEDIR` the generated cmake-config accounts for the user-supplied installation directories. If supporting GNUInstallDirs is out of scope for Vcpkg, I suppose this MR may be closed

How did you get output like this out of vcpkg?

**Comment by BillyONeal** [comment]
To clarify, I'm not sure what 'support' gnuinstalldirs means. What those are set to when we are building the vcpkg port are unrelated to what you set them to at consumption time.

**Comment by SomeoneSerge** [comment]
> To clarify, I'm not sure what 'support' gnuinstalldirs means. What those are set to when we are building the vcpkg port are unrelated to what you set them to at consumption time.

I set `CMAKE_INSTALL_INCLUDEDIR` at imgui build time to control where its headers are installed, and where they will be consumed from later

**Comment by SomeoneSerge** [comment]
> How did you get output like this out of vcpkg?

OK, this one has been slightly misleading. Just to keep this technically correct, you should get a similar output if you set `CMAKE_INSTALL_PREFIX` to `/aaaa-imgui-1.90.6` and `CMAKE_INSTALL_INCLUDEDIR` to `/bbbb-imgui-1.90.6-dev/include` (which would be ignored), and then forcefully moved (yes this sounds strange, bear with me) the headers from `/aaaa-imgui-1.90.6` to `/bbbb-imgui-1.90.6-dev/include`. The "non-existing path" message here has nothing to do with Vcpkg, only the fact that setting `CMAKE_INSTALL_INCLUDEDIR` has no effect does. To re-iterate, I do not wish to insist or to create any animosity. If you think this is out of scope or that I'm wasting your time, I'd rather just leave

**Comment by dg0yt** [comment]
Maybe adapting to `GNUInstallDirs` wouldn't be a problem but it isn't needed in vcpkg. The layout is fixed. (Actually, `GNUInstallDirs` can make problems with "special" prefixes.)

But installation "components" is something which isn't used here, isn't tested here, and thus shouldn't be maintained here. And AFAIU that's the subject of this PR: "allow split outputs". Vcpkg has no use for "split outputs" wrt components.

**Comment by SomeoneSerge** [comment]
Thank you @dg0yt, that's fair enough!

**Comment by BillyONeal** [comment]
Thanks for your contribution, and sorry we didn't end up agreeing to merge!

**Pull request closed without merging**